src/util/string.h

```diff
@@ -17,6 +18,55 @@
 #include <vector>
 
 namespace util {
+/** Type to denote a format string that was checked at compile time */
+template <unsigned num_params>
+struct ConstevalFormatString {
+    const char* const fmt;
+    consteval ConstevalFormatString(const char* str) : fmt{str} { Detail_CheckNumFormatSpecifiers(fmt); }
+    constexpr static void Detail_CheckNumFormatSpecifiers(std::string_view str)
+    {
+        unsigned count_normal{0}; // Number of "normal" specifiers, like %s
+        unsigned count_pos{0};    // Max number in positional specifier, like %8$s
+        for (auto it{str.begin()}; it < str.end();) {
+            if (*it != '%') {
+                ++it;
+                continue;
+            }
+
+            if (++it >= str.end()) throw "Format specifier incorrectly terminated by end of string";
+            if (*it == '%') {
+                // %%
+                ++it;
+                continue;
+            }
+
+            // Test for positional specifier, like %8$
+            auto it_num{it};
+            unsigned maybe_num{0};
+            while ('0' <= *it_num && *it_num <= '9') {
+                maybe_num *= 10;
+                maybe_num += *it_num - '0';
+                ++it_num;
+            };
+
+            if (*it_num == '$') {
+                // Positional specifier
+                if (maybe_num == 0) throw "Positional format specifier must indicate at least 1";
+                count_pos = std::max(count_pos, maybe_num);
```

Comment:
Alternatively, we could cheaply collect the referenced indexes to a `std::set<unsigned> referenced_params;` and check that all of them were referenced, i.e. `if (referenced_params.size() != num_params) throw "Not all parameters are referenced!";`